Tools/machines/polaris-alcf/install_gpu_dependencies.sh

```diff
@@ -0,0 +1,123 @@
+#
+SW_DIR="/home/${USER}/sw/polaris/gpu"
+rm -rf ${SW_DIR}
+mkdir -p ${SW_DIR}
+
+# remove common user mistakes in python, located in .local instead of a venv
+python3 -m pip uninstall -qq -y pywarpx
+python3 -m pip uninstall -qq -y warpx
+python3 -m pip uninstall -qqq -y mpi4py 2>/dev/null || true
+
+# General extra dependencies ##################################################
+#
+
+# c-blosc (I/O compression)
+if [ -d $HOME/src/c-blosc ]
+then
+  cd $HOME/src/c-blosc
+  git fetch --prune
+  git checkout v1.21.1
+  cd -
+else
+  git clone -b v1.21.1 https://github.com/Blosc/c-blosc.git $HOME/src/c-blosc
+fi
+rm -rf $HOME/src/c-blosc-pm-gpu-build
+cmake -S $HOME/src/c-blosc -B $HOME/src/c-blosc-pm-gpu-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=${SW_DIR}/c-blosc-1.21.1
+cmake --build $HOME/src/c-blosc-pm-gpu-build --target install --parallel 16
+rm -rf $HOME/src/c-blosc-pm-gpu-build
+
+# ADIOS2
+if [ -d $HOME/src/adios2 ]
+then
+  cd $HOME/src/adios2
+  git fetch --prune
+  git checkout v2.8.3
+  cd -
+else
+  git clone -b v2.8.3 https://github.com/ornladios/ADIOS2.git $HOME/src/adios2
+fi
+rm -rf $HOME/src/adios2-pm-gpu-build
+cmake -S $HOME/src/adios2 -B $HOME/src/adios2-pm-gpu-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DADIOS2_USE_ZeroMQ=OFF -DCMAKE_INSTALL_PREFIX=${SW_DIR}/adios2-2.8.3
+cmake --build $HOME/src/adios2-pm-gpu-build --target install -j 16
+rm -rf $HOME/src/adios2-pm-gpu-build
+
+# BLAS++ (for PSATD+RZ)
+if [ -d $HOME/src/blaspp ]
+then
+  cd $HOME/src/blaspp
+  git fetch --prune
+  git checkout master
+  git pull
+  cd -
+else
+  git clone https://github.com/icl-utk-edu/blaspp.git $HOME/src/blaspp
+fi
+rm -rf $HOME/src/blaspp-pm-gpu-build
+CXX=$(which CC) cmake -S $HOME/src/blaspp -B $HOME/src/blaspp-pm-gpu-build -Duse_openmp=OFF -Dgpu_backend=cuda -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=${SW_DIR}/blaspp-master
+cmake --build $HOME/src/blaspp-pm-gpu-build --target install --parallel 16
+rm -rf $HOME/src/blaspp-pm-gpu-build
+
+# LAPACK++ (for PSATD+RZ)
+if [ -d $HOME/src/lapackpp ]
+then
+  cd $HOME/src/lapackpp
+  git fetch --prune
+  git checkout master
+  git pull
+  cd -
+else
+  git clone https://github.com/icl-utk-edu/lapackpp.git $HOME/src/lapackpp
+fi
+rm -rf $HOME/src/lapackpp-pm-gpu-build
+CXX=$(which CC) CXXFLAGS="-DLAPACK_FORTRAN_ADD_" cmake -S $HOME/src/lapackpp -B $HOME/src/lapackpp-pm-gpu-build -DCMAKE_CXX_STANDARD=17 -Dbuild_tests=OFF -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=ON -DCMAKE_INSTALL_PREFIX=${SW_DIR}/lapackpp-master
+cmake --build $HOME/src/lapackpp-pm-gpu-build --target install --parallel 16
+rm -rf $HOME/src/lapackpp-pm-gpu-build
+
+# Python ######################################################################
+#
+python3 -m pip install --upgrade pip
+python3 -m pip install --upgrade virtualenv
+python3 -m pip cache purge
+rm -rf ${SW_DIR}/venvs/warpx
+python3 -m venv --system-site-packages ${SW_DIR}/venvs/warpx
```

Comment:
I'm needing to use `--system-site-packages` since I cannot manage to install `mpi4py` in a virtual environment.

Reply:
Can you share more details on the issue?

Reply:
Here's what I sent to ALCF support:

> I’m trying to install WarpX<https://github.com/ECP-WarpX/WarpX> with Python support on Polaris and as part of this I would like to create a new Python virtual environment in which to install all the required Python libraries. However, I’m having trouble getting `mpi4py` installed in the virtual environment and would appreciate any help you can give me.
> I’m currently performing the following steps:
>  
> 1. Loading required modules:
>  
> module load cmake/3.23.2
>  
> module load cudatoolkit-standalone
>  
> module load cray-python/3.9.13.1
>  
>  
> 1. Creating a virtual environment:
>  
> python3 -m venv venvs/warpx
>  
> source venvs/warpx/bin/activate
>  
> 1. Installing pre-requisite Python packages:
>  
> python3 -m pip install --upgrade pip
>  
> python3 -m pip install --upgrade build
>  
> python3 -m pip install --upgrade packaging
>  
> python3 -m pip install --upgrade wheel
>  
> python3 -m pip install --upgrade setuptools
>  
>  
> 1. Trying to install mpi4py:
>  
> python3 -m pip install mpi4py
>  
> The mpi4py installation fails when checking for required MPI libraries with messages of the form:
> checking for MPI compile and link ...
> /opt/cray/pe/mpich/8.1.25/ofi/nvidia/20.7/bin/mpicc -Wno-unused-result -Wsign-compare -DNDEBUG -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -DOPENSSL_LOAD_CONF -fwrapv -fno-semantic-interposition -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -fPIC -I/home/groenewa/scratch/venvs/sandbox/include -I/usr/include/python3.6m -c _configtest.c -o _configtest.o
> nvc-Error-Unknown switch: -Wno-unused-result
> nvc-Error-Unknown switch: -fmessage-length=0
> nvc-Error-Unknown switch: -grecord-gcc-switches
> nvc-Error-Unknown switch: -fstack-protector-strong
> nvc-Error-Unknown switch: -funwind-tables
> nvc-Error-Unknown switch: -fasynchronous-unwind-tables
> nvc-Error-Unknown switch: -fstack-clash-protection
> nvc-Error-Unknown switch: -fwrapv
> nvc-Error-Unknown switch: -fno-semantic-interposition
> nvc-Error-Unknown switch: -fmessage-length=0
> nvc-Error-Unknown switch: -grecord-gcc-switches
> nvc-Error-Unknown switch: -fstack-protector-strong
> nvc-Error-Unknown switch: -funwind-tables
> nvc-Error-Unknown switch: -fasynchronous-unwind-tables
> nvc-Error-Unknown switch: -fstack-clash-protection
> nvc-Error-Unknown switch: -fmessage-length=0
> nvc-Error-Unknown switch: -grecord-gcc-switches
> nvc-Error-Unknown switch: -fstack-protector-strong
> nvc-Error-Unknown switch: -funwind-tables
> nvc-Error-Unknown switch: -fasynchronous-unwind-tables
> nvc-Error-Unknown switch: -fstack-clash-protection
> failure.
> removing: _configtest.c _configtest.o
> error: Cannot compile MPI programs. Check your configuration!!!
> ERROR: Failed building wheel for mpi4py

They suggested:

> Try using the default conda module (conda/2022-09-08) instead of the cudatoolkit

I've been able to install `mpi4py` when loading the conda module and creating a virtual environment from it (not actually a conda virtual environment, just a `venv` virtual environment from the conda Python). I'm now just seeing if the rest of the build works fine.

Reply:
I can confirm now that the above problem goes away when not loading the `cudatoolkit` module.